### PR TITLE
[VL] Avoid throwing exception from destructor

### PR DESCRIPTION
**Pull request by jackylee-ch**
## What changes were proposed in this pull request?
`Terminated by ClosedByInterruptException` would be found while using speculation in Spark. The reason is that the `GlutenException` is not caught when the C++ object is destructed, which would cause the executor exit abnormally.

## How was this patch tested?
unit tests

### CODE REVIEW

**Comment by jackylee-ch** [comment]
It seems we meet the same problem with https://github.com/oap-project/gluten/pull/1505 in new code. @zhztheplayer @zhouyuan

**Review by zhouyuan — APPROVED** [review]
👍

**Comment by jackylee-ch** [comment]
Thanks for your review.

File: cpp/velox/jni/JniFileSystem.cc [hunk 1]
```diff
@@ -75,13 +75,13 @@ class JniReadFile : public facebook::velox::ReadFile {
   ~JniReadFile() override {
     try {
       close0();
+      JNIEnv* env;
+      attachCurrentThreadAsDaemonOrThrow(vm, &env);
+      env->DeleteGlobalRef(obj_);
+      checkException(env);
     } catch (const std::exception& e) {
       LOG(WARNING) << "Error closing jni read file " << e.what();
     }
```
Comment (zhztheplayer) [inline cpp/velox/jni/JniFileSystem.cc:84]: Thanks, and just out of my curiosity, do you actually use spill strategy `heap-over-local` ? @jackylee-ch

Comment (zhouyuan) [reply cpp/velox/jni/JniFileSystem.cc:84]: it's possibly due to speculation is enabled in their test, thus executors will be killed "randomly" (?)

thanks, -yuan

Comment (jackylee-ch) [reply cpp/velox/jni/JniFileSystem.cc:84]: > it's possibly due to speculation is enabled in their test, thus executors will be killed "randomly" (?)

@zhztheplayer @zhouyuan  The current problem is only related to speculation and I didn't use the spill strategy `heap-over-local`. 
As speculation is enabled, the task would be killed if another attempt success and would throw `ClosedByInterruptException` in JVM. The `checkException` would catch the exception and throw `GlutenException`, which was not catched by `try catch` and finally make executor failed.

Comment (zhztheplayer) [reply cpp/velox/jni/JniFileSystem.cc:84]: OK but this file `JniFileSystem.cc` should only be used when `heap-over-local` is enabled. Am I missing something here?

Comment (jackylee-ch) [reply cpp/velox/jni/JniFileSystem.cc:84]: https://github.com/oap-project/gluten/blob/main/cpp/velox/jni/VeloxJniWrapper.cc#L59C39-L59C39
Currently we would call `gluten::initVeloxJniFileSystem` in `JNI_OnLoad`.

Comment (jackylee-ch) [reply cpp/velox/jni/JniFileSystem.cc:84]: > OK but this file `JniFileSystem.cc` should only be used when `heap-over-local` is enabled. Am I missing something here?

@zhztheplayer Maybe we should add a check for the `initVeloxJniFileSystem`

**Pull request merged**